Add a new setting for whether the project is in limited or public view mode

pjanik opened the pull request:
This PR adds a new setting to a project: "public". It's true by default (migration handles that).
When a project isn't public, it's hidden in the following places:
- home page cards
- search filters
- nav bar collections menu
- material settings (you can't assign material to a hidden project)

However, you can still see hidden projects sometimes. In fact on all the pages under "Admin" tab. So, you can change hidden project settings or become its admin / researcher.

Also, if you're an admin or researcher for the project, you always see it. However "regular" admins don't see hidden projects in the places listed above. I assumed that "public = false" meant "work in progress", so I didn't want to pollute view for regular admins. Not sure about it though, just let me know, it's very easy to change.

scytacki commented:
Our convention is that a portal/regular admin has a superset of privileges. So it would be more consistent if portal/regular admins see these non public projects.

Instead of putting the `visible?` method in the model, it should go in the ProjectPolicy. Normally in the views you could do `policy(project).visible?`, but I don't see any places in this code where you should do that. I'm pretty sure in the Search model you'll need to use: `Pundit.policy!(user, project).visible?` because pundit won't be added automatically and there is no current_user available.

For the places with listings of projects it seems like you should use a Policy Scope. So this:
`Admin::Project.with_landing_pages.select { |p| p.visible?(current_visitor) }`
would be replaced with this
`policy_scope(Admin::Project).with_landing_pages`

Unfortunately that is complicated, because the Admin::Project scope is currently used to limit which projects show up in the projects listing to admins. There doesn't seem to be a nice way to handle multiple scopes like this (but perhaps I'm not thinking about it right).  
So it is probably best to stick with the select and do:
`Admin::Project.with_landing_pages.select { |p| policy(p).visible? }` in this case Pundit should be available and it should pick up the current_user.

pjanik commented:
Thanks for those tips, I hadn't learned more about Pundit before as I thought it's used only by controllers. Basically I've changed things according to your comments, but a few notes below.

>  I'm pretty sure in the Search model you'll need to use: Pundit.policy!(user, project).visible?

Actually, we use a wrapper around `User` object which is specified in `ApplicationController` and provides additional context (pretty specific to controller env: request, params, etc.). I would need to initialize this context object first, fill it with nils and finally call `Pundit.policy!(context, project).visible?`. I guess it would be pretty annoying long-term, so I modified `ApplicationPolicy`. Now it accepts either a context object or a plain `User` instance. More convenient and it doesn't seem to be any worse, as we should remember that additional context might be undefined no matter which approach we pick.

> There doesn't seem to be a nice way to handle multiple scopes like this (but perhaps I'm not thinking about it right).

"Recommended" approach:
```ruby
def index
  @posts = PostPolicy::Scope.new(current_user, Post).your_custom_scope_method
end
```
from https://github.com/elabs/pundit/issues/234,
but I wouldn't call it nice. I can't really see a big benefit here, except from the fact that we move the scope implementation to separate file. But scopes become pretty painful. While pundit policies seem to be reasonable, I feel that a single scope is a bit limited. 

Maybe we could assume that Pundit scope is a basic security pre-filter that should be always applied and then it can be further limited by controllers / views. So, theoretically we could use Pundit scope to limit visibility. The problem is that probably it's pretty hard to do it using queries, so the result would be an array and we wouldn't be able to chain other scopes. Also, not sure if project visibility is so crucial so it's worth it. It can make sense when we're concerned about privacy, security, etc. That's just my first impression though, I'm touching Pundit scopes for the first time.

I've also had problems with rspec. `policy` helper method was undefined in some view, shared examples. I was trying to google for an easy solution, found some threads but no clear answer. I've just removed project filtering in material settings, as now it's not necessary (project listing there is visible only for admins).

scytacki commented on app/models/user.rb:
```diff
@@ -591,6 +595,10 @@ def cohorts
   end
 
+  def cohort_projects
+    cohorts.map {|c| c.project}.flatten.compact.uniq
```
This might be better defined like this: `has_many :cohort_projects, through: :cohorts, source: :project, uniq: true`
I'm not sure I've got that right. But by doing it that it should be a bit more efficient because it ought to use a SQL to do the map/join, and uniq.

pjanik replied:
Hm, but `cohorts` is a method, not association. There is no direct association between `User` and `Cohort`, cohorts are inferred from `portal_teacher` or `portal_student` instances. Student's cohorts are actually inferred from his teachers' cohorts, teachers are inferred from classes etc., and most of that is implemented as methods, not associations. Maybe it's possible to change everything to associations. At the end we would need to define method anyway, like:
```ruby
def cohort_projects
  teacher_cohort_projects || student_cohort_projects
end
```
I can try.

pjanik replied:
Fixed in: https://github.com/concord-consortium/rigse/commit/b0f3c2d96c27fb574f4493dbbee4cada51b7e8f3

scytacki commented:
Other than the change to `has_many` this looks good to go.

scytacki commented:
I agree with the idea of using the Pundit Scopes as a security pre-filter. So in this case the current Scope for Projects ought to be revised since all users should be able to see all public projects. And members of projects should be able to see their non-public projects too.

You don't need to make that part of this PR though. I just wanted to describe what I think of when using it is a security pre-filter.   Ideally I think I'd like to be able to define conditions that can be used as either scopes or for checking individual items. So something like `edit?` could be used to efficiently get a list of resources that can be edited by the current user, or it would simply be used to check if the current user can edit the current item. Looks like CanCanCan provides this: https://github.com/CanCanCommunity/cancancan/wiki/Fetching-Records
The way Pundit does things it seems like it would never by possible to take this approach. Sigh.

scytacki commented on app/models/portal/student.rb:
```diff
@@ -16,6 +16,10 @@ class Portal::Student < ActiveRecord::Base
+  has_many :teachers, :through => :clazzes, :class_name => "Portal::Teacher", :source => :teachers, :uniq => true
+  # students cohorts are infered from its teacher(s)
+  has_many :cohorts, :through => :teachers, :class_name => "Admin::Cohort", :source => :cohorts, :uniq => true
+  has_many :projects, :through => :cohorts, :class_name => "Admin::Project", :source => :project, :uniq => true
```
Can you leave off `:source` key in these rows because AR should use the name of the relationship by default?  
I didn't know you could chain these :through relations like that, very cool.

pjanik replied:
I wasn't sure about plural and singular forms (e.g. `cohorts` => `cohorts` vs `projects` => `project`). It seems to work. :wink:

scytacki commented:
:+1: Looks good. I'm just going to wait for the tests to finish running and them we can merge it in. If it helps feel free to merge it yourself. I won't be getting back to this for another 4 hours or so.
